packages/components/src/custom-select-control-v2/types.ts

```diff
@@ -0,0 +1,58 @@
+/**
+ * External dependencies
+ */
+// eslint-disable-next-line no-restricted-imports
+import type * as Ariakit from '@ariakit/react';
+
+export type CustomSelectContext =
+	| {
+			/**
+			 * The store object returned by Ariakit's `useSelectStore` hook.
+			 */
+			store: Ariakit.SelectStore;
+	  }
+	| undefined;
+
+export type CustomSelectProps = {
+	/**
+	 * The child elements. This should be composed of CustomSelect.Item components.
+	 */
+	children: React.ReactNode;
+	/**
+	 * An optional default value for the control. If left `undefined`, the first
+	 * non-disabled item will be used.
+	 */
+	defaultValue?: string | string[];
+	/**
+	 * Label for the control.
+	 */
+	label?: string;
+	/**
+	 * A function that receives the new value of the input.
+	 */
+	onChange?: ( newValue: string | string[] ) => void;
+	/**
+	 * Can be used to render select UI with custom styled values.
+	 */
+	renderSelectedValue?: ( selectValue: string | string[] ) => React.ReactNode;
+	/**
+	 * The size of the control.
+	 */
+	size?: 'default' | 'large';
+	/**
+	 * Can be used to externally control the value of the control.
+	 */
+	value?: string | string[];
+};
+
+export type CustomSelectItemProps = {
+	/**
+	 * The children to display for each select item.
+	 */
+	children: React.ReactNode;
+	/**
+	 * The value of the select item. This will be used as the children if
+	 * children are left `undefined`.
+	 */
+	value?: string;
```

Comment:
Did we ever find out why a consumer of `CustomSelectItem` would ever not pass a `value` ? In case we haven't found a good reason, `value` should probably be a required prop.

Reply:
@diegohaz shared that it's inspired by the native select element. 

So TIL it's optional for a `<select/>`! 😄  [From MDN:](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/option#value
) 
>If this attribute is omitted, the value is taken from the text content of the option element.

An `<option>` could be added with no children or `value` attribute, and it would become a selectable empty option. 😅 

And if we wanted a title for a group (I believe we thought this might be a possible reason), [it looks like `<optgroup>` is meant for that](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/optgroup) / [`<SelectGroup/>` from Ariakit](https://ariakit.org/examples/select-group) (which would be worth exploring in a follow-up).  

To be flexible and to match the native select and Ariakit, it seems like we might want to have both `value` and `children` as optional? What do you think @ciampo?

Reply:
Thank you for looking into this!

> it would become a _selectable_ empty option.

I quickly tested adding an item with no `value` prop ([demo](https://stackblitz.com/edit/vbogr4?file=select%2Fstyle.css,select%2Findex.tsx&theme=dark)), and that item is _not selectable_. 
On top of that, in terms of backwards-compat, it's much easier to allow a mandatory prop to be optional, than the other way around.

For those two reasons, I think it's better to start with the `value` prop being mandatory. We can always make it optional at a later point if we feel the need for it.

Reply:
Yes, I had misunderstood @brookewp's question. The [`value`](https://ariakit.org/reference/select-item#value) prop on `SelectItem` is optional so you can render non-selectable items.

We have a use case for that on the Ariakit site in the version select widget. You can select the version, but the "Changelog" item performs a different action:
![Screenshot of the version select widget on Ariakit.org](https://github.com/WordPress/gutenberg/assets/3068563/51fa8d16-28e6-455d-8443-789c88bb237a)